Issues/427 add edit image meta *** DO NOT MERGE ***

aerych opened the pull request:
First pass (and work in progress) for adding the ability to edit image meta data, e.g. link, caption, etc.   Ready for your feedback!! 

### Points of interest!
- Adds two new JS libraries: underscore.js and shortcode.js (from core WordPress)
- Image parsing logic was borrowed and modified from wp core in order to be as consistent as possible with existing user expectations
- Adds a WPImageMeta convenience class as a simple holder for meta values. 
- Adds a simple modal to demo the edit feature. 

### TODOs
- There is some data sanitation and safety logic that needs to be implemented.
- There are some slight differences in the logic that parses out the image, and the logic that rebuilds it. These need to be made uniform. 
- Some meta values can be derived from other meta values so the total number can likely be reduced (by at least 2)

### Questions:
What are your thoughts on the best approach to the demo form? Its as simple as possible right now. 

### Misc:
I'm sure I'm forgetting something. :) 

@SergioEstevao @bummytime @diegoreymendez

aerych commented:
@SergioEstevao @bummytime @diegoreymendez 
Ready for another look :)

bummytime commented:
@aerych I think this is great. I will let @diegoreymendez chime in too, but it is a :shipit: from me!

diegoreymendez commented on Assets/ZSSRichTextEditor.js:
```diff
@@ -806,8 +806,446 @@ ZSSEditor.removeImage = function(imageNodeIdentifier) {
+ *  @param      imageMetaString   A JSON string representing the updated meta data.
+ */
+ZSSEditor.updateCurrentImageMeta = function( imageMetaString ) {
+    //    ZSSEditor.restoreRange();
```
This line should be removed or a comment added to explain its purpose.

aerych replied:
Good eye.

diegoreymendez commented on Assets/ZSSRichTextEditor.js:
```diff
@@ -806,8 +806,446 @@ ZSSEditor.removeImage = function(imageNodeIdentifier) {
+        return; // return to avoid sending enabled styles twice.
+    }
+
+    ZSSEditor.sendEnabledStyles();
```
Does this method need to send the enabled styles when it does nothing?

aerych replied:
Actually, I think it doesn't.

diegoreymendez commented on Assets/ZSSRichTextEditor.js:
```diff
@@ -806,8 +806,446 @@ ZSSEditor.removeImage = function(imageNodeIdentifier) {
+ */
+ZSSEditor.selectImageCaptionNode = function( imageNode ) {
+    var node = imageNode;
+    if ( node.parentNode && node.parentNode.nodeName === 'A' ) {
```
I'd consider changing the second expression to `nodeName.toLowerCase() == 'a'` to avoid having to deal with case sensitivity.

bummytime replied:
I always thought ```nodeName``` returned the name in uppercase. Better safe than sorry I suppose...

http://ejohn.org/blog/nodename-case-sensitivity/

TooTallNate replied:
`nodeName` is definitely always upper-case

aerych replied:
Going to leave this one as is since this is the same practice used in core, and also Nathans comment.

diegoreymendez replied:
@TooTallNate - Thanks for the tip.  We'll be able to revert the other `toLowerCase()` at some point, then.

diegoreymendez commented on Assets/ZSSRichTextEditor.js:
```diff
@@ -806,8 +806,446 @@ ZSSEditor.removeImage = function(imageNodeIdentifier) {
+    var node = imageNode;
+    if ( node.parentNode && node.parentNode.nodeName === 'A' ) {
+        node = node.parentNode;
+    }
```
Following the logic I was wondering if the `else` condition should halt execution of this method.  Same for the `if` blocks below.

aerych replied:
Each if block should be checked independently. An image may or may not have an anchor, or a caption. This is kind of explained in the brief but maybe it could be clarified.

diegoreymendez commented on Assets/ZSSRichTextEditor.js:
```diff
@@ -806,8 +806,446 @@ ZSSEditor.removeImage = function(imageNodeIdentifier) {
+    var node = imageNode;
+
+    // If the image's parent is an anchor, let the anchor be the working node.
+    if ( node.parentNode && node.parentNode.nodeName === 'A' ) {
```
Since this code is almost identical to the code written in `ZSSEditor.selectImageCaptionNode` maybe we can have a method to obtain the caption node (ie: `getCaptionNodeForImage` or something similar).

aerych replied:
Its a good thought :)  Will do

diegoreymendez commented on Classes/WPEditorView.m:
```diff
@@ -205,33 +206,28 @@ - (NSString*)editorRawHTML
 }
 
-- (NSString*)editorScript
+- (NSString *)javascriptFromBundleResourceNamed:(NSString *)filename
```
:+1: :two:

diegoreymendez commented:
@aerych - Just reviewed the code... I'll be testing this and will let you know if there's anything on that front.

aerych commented:
I've updated this branch with a better approach to inserting the formatted caption. There was some strange behavior with the content-editiable insertHTML method depending on what the adjacent markup was.  
I've also update the branch with selected styles similar to as discussed in #474. I've taken this a bit further and alter the selected appearance according to the size of the image.  The following screen shots demo the different levels (note that I just reuse the same image and force the different styles to show how they might look). 

Images that are large enough will show the full selection overlay, which is the circled icon plus the text.
If the image is too small to show the full overlay, the circle is dropped. 
A very small image will only show the icon.
@bummytime @diegoreymendez @SergioEstevao  Ready for yet another round!

### large image
![ios simulator screen shot jan 19 2015 6 13 55 pm](https://cloud.githubusercontent.com/assets/1435271/5810201/eb801442-a007-11e4-8f59-3a56a4b3adf7.png)


### smaller image
![ios simulator screen shot jan 19 2015 6 14 54 pm](https://cloud.githubusercontent.com/assets/1435271/5810202/f19108aa-a007-11e4-8d41-f1ae55a50786.png)


### small image
![ios simulator screen shot jan 19 2015 6 15 31 pm](https://cloud.githubusercontent.com/assets/1435271/5810203/f7df7322-a007-11e4-862e-3ee07ce48484.png)

bummytime commented:
Hola @aerych. Nice work improving this feature. I have not started reviewing the latest code yet, but I did run through the feature itself and had a few general comments. First thing I noticed is that on smaller screens (iPhone 4s), the edit icon appears very large:

![screen shot 2015-01-19 at 7 51 12 pm](https://cloud.githubusercontent.com/assets/154014/5811207/a8c0e08a-a018-11e4-93d2-7d9a92b74a9e.png)

Is this something that could get scaled down? For reference, it looks good on iPads:

![screen shot 2015-01-19 at 8 23 05 pm](https://cloud.githubusercontent.com/assets/154014/5811228/0e6710d0-a019-11e4-8f95-92016e4b5ca8.png)

Next I made a quick video demonstrating 3 potential issues:

https://cloudup.com/cQiYu40qpy4

1) When the image is "uploading", I can still tap it and see the edit icon. In addition it appears to interfere with the behavior @SergioEstevao introduced with upload progress (tap to stop an upload). 

2) Note at the end of the video there is a weird issue where the progress bar never disappears after tapping edit during an upload. 

3) The content view seems to jump around a bit when editing. Is there anything we can do to combat UIWebView here?

For points 1 and 2, maybe a quick solution is to disable image editing during an upload OR if the image is in a failed state?

Moving on to code next...

bummytime commented:
@aerych Also I noticed the image edges were fuzzy again. Was this on purpose?

bummytime commented on Assets/editor.html:
```diff
@@ -930,7 +933,66 @@
+
+        .edit-container img {
+            -webkit-filter: blur(4px) grayscale(0.3);
+            margin:-0.1px; /*tiny margin to keep crisp edges when blurring the image*/
```
Does this need to be a larger value for crisper edges?

aerych commented:
> First thing I noticed is that on smaller screens (iPhone 4s), the edit icon appears very large:

Should we we scale down the graphic or adjust the points at which the different parts of the overlay are taken away? 

> For points 1 and 2, maybe a quick solution is to disable image editing during an upload OR if the image is in a failed state?

Good catch re the conflict with uploading. My original intent was to not show the editor while uploading. I'll make the change.

> Also I noticed the image edges were fuzzy again. Was this on purpose?

No.  I'm thinking the css probably needs tweaking to account for different screen/image sizes.

aerych commented:
Hiya @bummytime. 

I've added a check for an uploading image and adjusted the size threshold for the large style overlay. 

I'm wondering if the fuzziness might be a simulator issue.  I can reproduce it (kind of) in the iPad simulator where the majority of the image edge is crisp but a portion is fuzzy.  I can't reproduce it on an actual device.  Does this match your experience? 

Re: the scrolling.  We touched on this in chat and I followed up with a test in the current develop branch. I can reproduce the scrolling behavior there by tapping on text or an image in the editor.  I think addressing that behavior is better handled in its own PR since its preexisting and this PR is already rather long.

bummytime commented:
Looks great @aerych :shipit: this bad boy!

aerych commented:
Awesome, thanks @bummytime!  Closing this PR in favor of $490 as it targets the correct release branch. Will merge that one.
